File: sdk/python/cortex_rmvm_sdk/plan.py
from __future__ import annotations

from .types import (
    ApplySelectorOp,
    AssertOp,
    FilterOp,
    JoinOp,
    Manifest,
    PlanInput,
    ProjectOp,
    ResolveOp,
)


def build_plan_only_prompt(user_message: str, manifest: Manifest) -> str:
    handles = ", ".join(h.ref for h in manifest.handles)
    selectors = ", ".join(manifest.selector_refs)
    return "\n".join(
        [
            "Return plan JSON only. Do not include prose.",
            "Generate a valid RMVMPlan with requestId, steps, outputs.",
            f"User message: {user_message}",
            f"Allowed handle refs: [{handles}]",
            f"Allowed selector refs: [{selectors}]",
            "Every fetch.handleRef must be from allowed handle refs.",
            "Every applySelector.selectorRef must be from allowed selector refs.",
        ]
    )


def validate_plan_against_manifest(plan: PlanInput, manifest: Manifest) -> None:
    handle_refs = {h.ref for h in manifest.handles}
    selector_refs = set(manifest.selector_refs)
    regs: set[str] = set()

    for step in plan.steps:
        if not step.out:
            raise ValueError("invalid plan: step.out is required")
        if step.out in regs:
            raise ValueError(f"invalid plan: register redefined ({step.out})")

        op = step.op
        if hasattr(op, "handle_ref"):
            if op.handle_ref not in handle_refs:
                raise ValueError(f"invalid plan: unknown handle ref {op.handle_ref}")
        elif isinstance(op, ApplySelectorOp):
            if op.selector_ref not in selector_refs:
                raise ValueError(f"invalid plan: unknown selector ref {op.selector_ref}")
        elif isinstance(op, (ResolveOp, FilterOp, ProjectOp)):
            if op.in_reg not in regs:
                raise ValueError(f"invalid plan: input register not defined ({op.in_reg})")
        elif isinstance(op, JoinOp):
            if op.left_reg not in regs or op.right_reg not in regs:
                raise ValueError("invalid plan: join registers not defined")
        elif isinstance(op, AssertOp):
            for binding in op.bindings.values():
                if binding.reg not in regs:
                    raise ValueError(
                        f"invalid plan: assert binding register not defined ({binding.reg})"
                    )

        regs.add(step.out)

File: sdk/typescript/src/plan.ts
import type { Manifest, PlanInput } from "./types.js";

export function buildPlanOnlyPrompt(userMessage: string, manifest: Manifest): string {
  const allowedHandles = manifest.handles.map((h) => h.ref).join(", ");
  const allowedSelectors = manifest.selectorRefs.join(", ");
  return [
    "Return plan JSON only. Do not include prose.",
    "Generate a valid RMVMPlan with requestId, steps, outputs.",
    `User message: ${userMessage}`,
    `Allowed handle refs: [${allowedHandles}]`,
    `Allowed selector refs: [${allowedSelectors}]`,
    "Every fetch.handleRef must be from allowed handle refs.",
    "Every applySelector.selectorRef must be from allowed selector refs.",
  ].join("\n");
}

export function validatePlanAgainstManifest(plan: PlanInput, manifest: Manifest): void {
  const handleRefs = new Set(manifest.handles.map((h) => h.ref));
  const selectorRefs = new Set(manifest.selectorRefs);
  const regs = new Set<string>();

  for (const step of plan.steps) {
    if (!step.out) {
      throw new Error("invalid plan: step.out is required");
    }
    if (regs.has(step.out)) {
      throw new Error(`invalid plan: register redefined (${step.out})`);
    }

    switch (step.op.kind) {
      case "fetch":
        if (!handleRefs.has(step.op.handleRef)) {
          throw new Error(`invalid plan: unknown handle ref ${step.op.handleRef}`);
        }
        break;
      case "applySelector":
        if (!selectorRefs.has(step.op.selectorRef)) {
          throw new Error(`invalid plan: unknown selector ref ${step.op.selectorRef}`);
        }
        break;
      case "resolve":
      case "filter":
      case "project":
        if (!regs.has(step.op.inReg)) {
          throw new Error(`invalid plan: input register not defined (${step.op.inReg})`);
        }
        break;
      case "join":
        if (!regs.has(step.op.leftReg) || !regs.has(step.op.rightReg)) {
          throw new Error("invalid plan: join registers not defined");
        }
        break;
      case "assert":
        for (const binding of Object.values(step.op.bindings)) {
          if (!regs.has(binding.reg)) {
            throw new Error(`invalid plan: assert binding register not defined (${binding.reg})`);
          }
        }
        break;
      default:
        break;
    }

    regs.add(step.out);
  }
}
